File: Daily Test/C++/Maximum Sum SubArray - Common Digits 3.11.2025.cpp
#include <iostream>
#include <vector>
#include <set>
#include <string>
using namespace std;

set<char> digitset(int x) {
    set<char> s;
    string str = to_string(x);
    for (char c : str) s.insert(c);
    return s;
}

int main() {
    int n;
    cin >> n;
    vector<int> a(n);
    for (int& x : a) cin >> x;
    int bl = 0, bs = 0;
    for (int i = 0; i < n; ++i) {
        set<char> c = digitset(a[i]);
        int s = a[i];
        for (int j = i + 1; j < n; ++j) {
            set<char> next = digitset(a[j]);
            set<char> inter;
            for (char d : c) if (next.count(d)) inter.insert(d);
            c = inter;
            if (c.empty()) break;
            s += a[j];
            int l = j - i + 1;
            if (l > bl || (l == bl && s > bs)) {
                bl = l;
                bs = s;
            }
        }
    }
    if (bl == 0) {
        bs = a[0];
        for (int i = 1; i < n; i++) if (a[i] > bs) bs = a[i];
    }
    cout << bs << endl;
    return 0;
}

File: Daily Test/Java/Maximum Sum SubArray - Common Digits 3.11.2025.java
import java.util.*;

public class Main {
    static Set<Character> digitSet(int x) {
        Set<Character> s = new HashSet<>();
        for (char c : Integer.toString(x).toCharArray()) s.add(c);
        return s;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] a = new int[n];
        for (int i = 0; i < n; i++) a[i] = sc.nextInt();
        int bl = 0, bs = 0;
        for (int i = 0; i < n; i++) {
            Set<Character> c = digitSet(a[i]);
            int s = a[i];
            for (int j = i + 1; j < n; j++) {
                Set<Character> next = digitSet(a[j]);
                c.retainAll(next);
                if (c.isEmpty()) break;
                s += a[j];
                int l = j - i + 1;
                if (l > bl || (l == bl && s > bs)) {
                    bl = l;
                    bs = s;
                }
            }
        }
        if (bl == 0) {
            bs = a[0];
            for (int i = 1; i < n; i++) if (a[i] > bs) bs = a[i];
        }
        System.out.println(bs);
        sc.close();
    }
}